### PR TITLE
Grizzly-http and grizzly-client instrumentation

### PR DESCRIPTION
**Pull request by heathkd**
The Mulesoft instrumentation traces API calls made from HttpListener and HttpRequester components. Under the hood, these components use grizzly for both the server and client logic. As a result, this PR adds instrumentation to Grizzly. 

For development purposes, the instrumentation is currently located in the mule-http-connector module of the tracer. However, it should be added to the grizzly module before being merged.

### CODE REVIEW

**Comment by heathkd** [comment]
The instrumentation in the "clientV2" package is some experimental instrumentation where the client span is created from `HttpClientFilter.java` instead of `AsyncHttpClient.java`. However, I'd like to focus on using the instrumentation in the `client` packaged and the `server` package. 

Currently, the client and server span are being sent as separate traces. I would like to consolidate the two spans into one trace. If there is anything obvious that is missing and should be added to `ClientAdvice.java` or `ServerAdvice.java` to make this happen, please let me know.

File: dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/AbstractExecutorInstrumentation.java [hunk 1]
```diff
@@ -81,7 +81,11 @@ public AbstractExecutorInstrumentation(final String... additionalNames) {
         "kotlinx.coroutines.scheduling.CoroutineScheduler",
         "org.eclipse.jetty.util.thread.QueuedThreadPool",
         "org.eclipse.jetty.util.thread.ReservedThreadExecutor",
+        "org.glassfish.grizzly.threadpool.AbstractThreadPool",
+        "org.glassfish.grizzly.threadpool.FixedThreadPool",
         "org.glassfish.grizzly.threadpool.GrizzlyExecutorService",
+        "org.glassfish.grizzly.threadpool.QueueLimitedThreadPool",
+        "org.glassfish.grizzly.threadpool.SyncThreadPool",
```
Comment (heathkd) [inline dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/AbstractExecutorInstrumentation.java:88]: Executor instrumentation was skipped for these four executors. FixedThreadPool, specifically, is instantiated by GrizzlyExecutorService after HttpServerFilter receives a request. QueueLimitedThreadPool is created if a FixedThreadPool cannot be created. Fixed, QueueLimited, and Sync all extend AbstractThreadPool. Would it make more sense to just add AbstractThreadPool to the whitelist?

File: dd-java-agent/instrumentation/mule-http-connector-1.0/src/main/java/datadog/trace/instrumentation/mulehttpconnector/client/ClientRequestAdvice.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    final AgentSpan span = startSpan("http.request");
+    final AgentScope scope = activateSpan(span, false);
+
+    DECORATE.afterStart(span);
+    DECORATE.onRequest(span, request);
+    propagate().inject(span, request, SETTER);
+
+    InstrumentationContext.get(AsyncCompletionHandler.class, AgentSpan.class)
+        .put((AsyncCompletionHandler) handler, span);
+    scope.close();
```
Comment (heathkd) [inline dd-java-agent/instrumentation/mule-http-connector-1.0/src/main/java/datadog/trace/instrumentation/mulehttpconnector/client/ClientRequestAdvice.java:37]: Does the span need to be activated here? Should a scope be created considering it will be closed at the end of the function.

File: dd-java-agent/instrumentation/mule-http-connector-1.0/src/main/java/datadog/trace/instrumentation/mulehttpconnector/client/ClientResponseAdvice.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+  public static void onExit(
+      @Advice.This final AsyncCompletionHandler handler,
+      @Advice.Argument(0) final Response response,
+      @Advice.Thrown final Throwable throwable) {
+
+    final ContextStore<AsyncCompletionHandler, AgentSpan> contextStore =
+        InstrumentationContext.get(AsyncCompletionHandler.class, AgentSpan.class);
+    final AgentSpan span = contextStore.get(handler);
+    final AgentScope scope = activateSpan(span, true);
```
Comment (heathkd) [inline dd-java-agent/instrumentation/mule-http-connector-1.0/src/main/java/datadog/trace/instrumentation/mulehttpconnector/client/ClientResponseAdvice.java:25]: This is in-line with the comment in ClientRequestAdvice. Should the span be reactivated here?

File: dd-java-agent/instrumentation/mule-http-connector-1.0/src/main/java/datadog/trace/instrumentation/mulehttpconnector/clientV2/ClientV2Decorator.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
+package datadog.trace.instrumentation.mulehttpconnector.clientV2;
```
Comment (heathkd) [inline dd-java-agent/instrumentation/mule-http-connector-1.0/src/main/java/datadog/trace/instrumentation/mulehttpconnector/clientV2/ClientV2Decorator.java:1]: Classes in the "clientV2" package can be ignored during review. Please review the classes in the "client" package instead. I was experimenting with slightly different instrumentation for the client.

File: dd-java-agent/instrumentation/mule-http-connector-1.0/src/main/java/datadog/trace/instrumentation/mulehttpconnector/clientV2/ClientV2Instrumentation.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
+package datadog.trace.instrumentation.mulehttpconnector.clientV2;
```
Comment (heathkd) [inline dd-java-agent/instrumentation/mule-http-connector-1.0/src/main/java/datadog/trace/instrumentation/mulehttpconnector/clientV2/ClientV2Instrumentation.java:1]: Classes in the "clientV2" package can be ignored during review. Please review the classes in the "client" package instead. I was experimenting with slightly different instrumentation for the client.

File: dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/AbstractExecutorInstrumentation.java (hunk 1 above)
Comment (heathkd) [reply dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/AbstractExecutorInstrumentation.java:88]: These executors were added to the whitelist to help resolve this issue: when a request is made to the server that then triggers a network request from the client, two traces are created instead of a single trace with the server span as the parent and the client span as the child. However, the problem still persists. Do I need to check that no executors related to the client have been skipped?

File: dd-java-agent/instrumentation/mule-http-connector-1.0/src/main/java/datadog/trace/instrumentation/mulehttpconnector/server/ServerAdvice.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    if (httpHeader instanceof HttpRequestPacket) {
+      final HttpRequestPacket httpRequest = (HttpRequestPacket) httpHeader;
+      final HttpResponsePacket httpResponse = httpRequest.getResponse();
+
+      final AgentSpan.Context parentContext = propagate().extract(httpRequest, GETTER);
+      final AgentSpan span = startSpan("mule.http.server", parentContext);
+
+      final AgentScope scope = activateSpan(span, false);
+      scope.setAsyncPropagation(true);
```
Comment (heathkd) [inline dd-java-agent/instrumentation/mule-http-connector-1.0/src/main/java/datadog/trace/instrumentation/mulehttpconnector/server/ServerAdvice.java:39]: When a request is made to the server that then triggers a network request from the client, two traces are created instead of a single trace with the server span as the parent and the client span as the child. I have set asyncPropagation to true, but I have also closed the scope. Would that result in the disconnected server and client spans?

File: dd-java-agent/instrumentation/mule-http-connector-1.0/src/test/groovy/MuleHttpServerTest.groovy [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        final HttpContent httpContent = ctx.getMessage();
+        final HttpRequestPacket request = (HttpRequestPacket) httpContent.getHttpHeader();
+
+        final HttpResponsePacket response = buildResponse(request)
+        ctx.write(response);
+      }
+      return ctx.getInvokeAction();
+    }
+
+    public HttpResponsePacket buildResponse(HttpRequestPacket request) {
```
Comment (heathkd) [inline dd-java-agent/instrumentation/mule-http-connector-1.0/src/test/groovy/MuleHttpServerTest.groovy:144]: This needs to be updated to call the `connector` method. It should also set `SUCCESS.body` to the body of the response instead of the reasonPhrase. However, it is unclear how to do this. It may be possible to write the response body using ctx.write().

File: dd-java-agent/instrumentation/mule-http-connector-1.0/src/main/java/datadog/trace/instrumentation/mulehttpconnector/filterchain/FilterchainAdvice.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+public class FilterchainAdvice {
+  @Advice.OnMethodEnter(suppress = Throwable.class)
+  public static AgentScope onEnter(@Advice.Argument(0) final FilterChainContext ctx) {
+
+    if (ctx.getAttributes().getAttribute(SPAN) == null || activeScope() != null) {
+      return null;
+    }
+
+    final AgentScope scope =
+        activateSpan((AgentSpan) ctx.getAttributes().getAttribute(SPAN), false);
```
Comment (heathkd) [inline dd-java-agent/instrumentation/mule-http-connector-1.0/src/main/java/datadog/trace/instrumentation/mulehttpconnector/filterchain/FilterchainAdvice.java:21]: The span is activated and the scope is explicitly closed. However, the logs point to an error where the tracer is attempting to close a specific scope while the current scope is "null". This code may be the cause for that error.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/api/ParentChildSpan.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
+package datadog.trace.bootstrap.instrumentation.api;
```
Comment (richardstartin) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/api/ParentChildSpan.java:1]: @tylerbenson FYI (ensures the class is loaded by the bootstrap classloader so doesn't require different inst points to share a classloader to coordinate via `InstrumentationContext`). Please ping me about whether this should really be necessary.

File: settings.gradle [hunk 10]
```diff
@@ -117,6 +117,7 @@ include ':dd-java-agent:instrumentation:log4j2'
 include ':dd-java-agent:instrumentation:mongo'
 include ':dd-java-agent:instrumentation:mongo:driver-3.1'
 include ':dd-java-agent:instrumentation:mongo:driver-async-3.3'
+include ':dd-java-agent:instrumentation:mule-http-connector-1.0'
```
Comment (tylerbenson) [inline settings.gradle:120]: Since this instrumentation all belongs to grizzly, let's move this to the `grizzly-2` folder.  The only `mule` reference should be in muzzle (which we decided we couldn't do), or a test, but I don't think we have any tests actually exercising mule, only grizzly.

File: dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/base/HttpServerTest.groovy [hunk 11]
```diff
@@ -431,6 +435,7 @@ abstract class HttpServerTest<SERVER> extends AgentTestRunner {
 
   def "test exception"() {
     setup:
+    assumeTrue(testException())
```
Comment (tylerbenson) [inline dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/base/HttpServerTest.groovy:438]: can you overload `testExceptionBody()` instead?

File: dd-java-agent/instrumentation/mule-http-connector-1.0/mule-http-connector-1.0.gradle [hunk 12]
```diff
@@ -0,0 +1,30 @@
+// no muzzle until/unless muzzle supports maven classifiers
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/mule-http-connector-1.0/mule-http-connector-1.0.gradle:1]: Please revert the change to the `integrations-core` submodule.

File: dd-java-agent/instrumentation/mule-http-connector-1.0/src/main/java/datadog/trace/instrumentation/mulehttpconnector/ContextAttributes.java [hunk 13]
```diff
@@ -0,0 +1,6 @@
+package datadog.trace.instrumentation.mulehttpconnector;
+
+public class ContextAttributes {
+  public static final String SPAN = "SPAN";
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/mule-http-connector-1.0/src/main/java/datadog/trace/instrumentation/mulehttpconnector/ContextAttributes.java:4]: See also: `HttpServerDecorator.DD_SPAN_ATTRIBUTE`

File: dd-java-agent/instrumentation/mule-http-connector-1.0/src/main/java/datadog/trace/instrumentation/mulehttpconnector/client/ClientRequestAdvice.java [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+    DECORATE.onRequest(span, request);
+    propagate().inject(span, request, SETTER);
+    InstrumentationContext.get(AsyncCompletionHandler.class, Pair.class)
+        .put((AsyncCompletionHandler<?>) handler, Pair.of(parentSpan, span));
+    return activateSpan(span);
+  }
+
+  @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+  public static void onExit(@Advice.Enter final AgentScope scope) {
+    scope.close();
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/mule-http-connector-1.0/src/main/java/datadog/trace/instrumentation/mulehttpconnector/client/ClientRequestAdvice.java:37]: Please add a comment that the scope is closed in `ClientResponseAdvice`

File: dd-java-agent/instrumentation/mule-http-connector-1.0/mule-http-connector-1.0.gradle [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
+  compileOnly group: 'org.mule.connectors', name: 'mule-http-connector', version: '1.5.14', classifier: 'mule-plugin'
+  compileOnly group: 'org.mule.runtime', name: 'mule-service-http-api', version: '4.1.1'
+  compileOnly group: 'org.mule.services', name: 'mule-service-http', version: '1.1.1'
+  compileOnly group: 'org.glassfish.grizzly', name: 'grizzly-http', version: '2.3.35'
+  compileOnly group: 'org.glassfish.grizzly', name: 'grizzly-http-client', version: '1.14'
+
+  testCompile group: 'org.glassfish.grizzly', name: 'grizzly-http-client', version: '1.14'
+  testCompile group: 'org.glassfish.grizzly', name: 'grizzly-http-server', version: '2.0'
+  testCompile project(':dd-java-agent:instrumentation:java-concurrent')
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/mule-http-connector-1.0/mule-http-connector-1.0.gradle:24]: For fun can you add a test dependency on `netty-3.8` and see if that has any impact when you run the client test?  `Netty38ClientTest` uses "ning" `async-http-client` which might already cover what we are doing here. (unless grizzly is using the same framework but swapping out the networking layer somehow, which would be interesting...)

File: dd-java-agent/instrumentation/mule-http-connector-1.0/src/test/groovy/GrizzlyAsyncHttpClientTest.groovy [hunk 16]
```diff
@@ -0,0 +1,68 @@
+import com.ning.http.client.*
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/mule-http-connector-1.0/src/test/groovy/GrizzlyAsyncHttpClientTest.groovy:1]: Please increase your wildcard limit in your IDE...

File: settings.gradle (hunk 10 above)
Comment (richardstartin) [reply settings.gradle:120]: I tried this just now, and they don't play nicely together. I expect it will be a decent sized piece of work to rationalise this.

File: dd-java-agent/instrumentation/mule-http-connector-1.0/mule-http-connector-1.0.gradle (hunk 15 above)
Comment (richardstartin) [reply dd-java-agent/instrumentation/mule-http-connector-1.0/mule-http-connector-1.0.gradle:24]: no impact at all.

Comment (richardstartin) [reply dd-java-agent/instrumentation/mule-http-connector-1.0/mule-http-connector-1.0.gradle:24]: It also has no impact if the dependency on `java-concurrent` is removed.

File: dd-java-agent/instrumentation/mule-http-connector-1.0/src/test/groovy/GrizzlyFilterchainServerTest.groovy [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+  }
+
+  @Override
+  String component() {
+    return ServerDecorator.DECORATE.component()
+  }
+
+  @Override
+  String expectedOperationName() {
+    return "http.request"
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/mule-http-connector-1.0/src/test/groovy/GrizzlyFilterchainServerTest.groovy:63]: Lets have this named the same as the other grizzly instrumentation...

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/api/Pair.java [hunk 18]
```diff
@@ -0,0 +1,32 @@
+package datadog.trace.bootstrap.instrumentation.api;
+
+public class Pair<T, U> {
```
Comment (dougqh) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/api/Pair.java:3]: final?

File: dd-java-agent/instrumentation/mule-http-connector-1.0/src/test/groovy/GrizzlyFilterchainServerTest.groovy (hunk 17 above)
Comment (dougqh) [reply dd-java-agent/instrumentation/mule-http-connector-1.0/src/test/groovy/GrizzlyFilterchainServerTest.groovy:63]: I think this depends a bit on the details.

As is, we're targeting Grizzly with no constraint on Mulesoft being present, so I think that makes sense.
However, if we want to restrict the activation, then there might be argument for handling this differently.

File: settings.gradle (hunk 10 above)
Comment (dougqh) [reply settings.gradle:120]: Yes, I understand the sentiment, but I'm not sure merging this with the other grizzly instrumentation is a good idea at this juncture.  

I think it would be good to restrict the activation to when Mulesoft is present, but for now, I think keeping this Mule specific makes sense.  The main question for me is will there be any adverse interactions.

File: dd-java-agent/instrumentation/mule-http-connector-1.0/src/test/groovy/GrizzlyFilterchainServerTest.groovy (hunk 17 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/mule-http-connector-1.0/src/test/groovy/GrizzlyFilterchainServerTest.groovy:63]: `http.request` just happens to be what most of our http client spans are named (where our server spans are generally named different), so in the absence of specific requirements I'm at least trying to be consistent.

Comment (richardstartin) [reply dd-java-agent/instrumentation/mule-http-connector-1.0/src/test/groovy/GrizzlyFilterchainServerTest.groovy]: @tylerbenson I have addressed this concern, making this instrumentation consistent with others. I have also created constants in `DDSpanNames` which both makes it easier to stay consistent but easier to get these common names into the interned UTF8 constants table.

File: settings.gradle (hunk 10 above)
Comment (richardstartin) [reply settings.gradle:120]: Can we agree on what to do here? The review suggestion changes the scope of the work considerably.

File: dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/base/HttpServerTest.groovy (hunk 11 above)
Comment (richardstartin) [reply dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/base/HttpServerTest.groovy:438]: They have different meaning.

File: dd-java-agent/instrumentation/grizzly-client-1.1/grizzly-client-1.1.gradle [hunk 19]
```diff
@@ -0,0 +1,30 @@
+// Set properties before any plugins get loaded
+ext {
+  minJavaVersionForTests = JavaVersion.VERSION_1_8
+}
+
+muzzle {
+  pass {
+    group = "org.glassfish.grizzly"
+    module = "grizzly-http-client"
```
[the remark this answers is not in the document]
Comment (richardstartin) [reply dd-java-agent/instrumentation/grizzly-client-1.1/grizzly-client-1.1.gradle]: addressed

File: dd-java-agent/instrumentation/grizzly-client-1.1/src/main/java/datadog/trace/instrumentation/grizzly/client/ClientRequestInstrumentation.java [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+import java.util.Map;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+@AutoService(Instrumenter.class)
+public final class ClientRequestInstrumentation extends Instrumenter.Default {
+
+  public ClientRequestInstrumentation() {
+    super("grizzly-client-1");
```
[the remark this answers is not in the document]
Comment (richardstartin) [reply dd-java-agent/instrumentation/grizzly-client-1.1/src/main/java/datadog/trace/instrumentation/grizzly/client/ClientRequestInstrumentation.java]: addressed

File: dd-java-agent/instrumentation/grizzly-client-1.1/src/main/java/datadog/trace/instrumentation/grizzly/client/ClientDecorator.java [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+import java.net.URI;
+import java.net.URISyntaxException;
+
+public class ClientDecorator extends HttpClientDecorator<Request, Response> {
+
+  public static final ClientDecorator DECORATE = new ClientDecorator();
+
+  @Override
+  protected String[] instrumentationNames() {
+    return new String[] {"grizzly-client-1"};
```
[the remark this answers is not in the document]
Comment (richardstartin) [reply dd-java-agent/instrumentation/grizzly-client-1.1/src/main/java/datadog/trace/instrumentation/grizzly/client/ClientDecorator.java]: addressed

File: dd-java-agent/instrumentation/grizzly-client-1.1/grizzly-client-1.1.gradle [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+    versions = "[1.9,1.16]"
```
[the remark this answers is not in the document]
Comment (richardstartin) [reply dd-java-agent/instrumentation/grizzly-client-1.1/grizzly-client-1.1.gradle]: addressed

File: settings.gradle [hunk 23]
```diff
@@ -117,6 +118,7 @@ include ':dd-java-agent:instrumentation:log4j2'
 include ':dd-java-agent:instrumentation:mongo'
 include ':dd-java-agent:instrumentation:mongo:driver-3.1'
 include ':dd-java-agent:instrumentation:mongo:driver-async-3.3'
+include ':dd-java-agent:instrumentation:grizzly-http-2.3.20'
```
Comment (tylerbenson) [inline settings.gradle:121]: Please alphabetize.

File: dd-trace-core/src/main/java/datadog/trace/core/StringTables.java [hunk 24]
```diff
@@ -36,6 +37,7 @@
     internConstantsUTF8(Tags.class);
     internConstantsUTF8(DDSpanTypes.class);
     internConstantsUTF8(DDComponents.class);
+    internConstantsUTF8(DDSpanNames.class);
```
Comment (devinsba) [inline dd-trace-core/src/main/java/datadog/trace/core/StringTables.java:40]: We might want a follow-up PR that goes through all the others and moves them here

Comment (richardstartin) [reply dd-trace-core/src/main/java/datadog/trace/core/StringTables.java:40]: fill your boots :)

**Comment by tylerbenson** [comment]
sorry, that was an accident.

File: dd-java-agent/instrumentation/grizzly-client-1.9/grizzly-client-1.9.gradle [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+
+testSets {
+  latestDepTest {
+    dirName = 'test'
+  }
+}
+
+dependencies {
+  compileOnly group: 'org.glassfish.grizzly', name: 'grizzly-http-client', version: '1.12'
+  testCompile group: 'org.glassfish.grizzly', name: 'grizzly-http-client', version: '1.12'
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/grizzly-client-1.9/grizzly-client-1.9.gradle:32]: Please set the compileOnly to 1.9 and add a comment saying that the tests only seem to work with 1.12.

Comment (richardstartin) [reply dd-java-agent/instrumentation/grizzly-client-1.9/grizzly-client-1.9.gradle]: Done. Would be good to get to the bottom of this one.
